osu.Game/Lists/IntervalList.cs

```diff
@@ -0,0 +1,128 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using osu.Framework.Lists;
+
+namespace osu.Game.Lists
+{
+    /// <summary>
+    /// Represents a list of intervals that can be used for whether a specific value falls into one of them.
+    /// </summary>
+    /// <typeparam name="T">The type of interval values.</typeparam>
+    public class IntervalList<T> : ICollection<Interval<T>>, IReadOnlyList<Interval<T>>
+    {
+        private static readonly IComparer<T> type_comparer = Comparer<T>.Default;
+
+        private readonly SortedList<Interval<T>> intervals = new SortedList<Interval<T>>((x, y) => type_comparer.Compare(x.Start, y.Start));
+
+        /// <summary>
+        /// The index of the nearest interval from last <see cref="IsInAnyInterval"/> call.
+        /// </summary>
+        protected int NearestIntervalIndex;
```

Comment:
why `protected`?

Reply:
I don't think we should expose such an internal field as `NearestIntervalIndex`, it's used for the search inside the method only, changed to `protected` for testing what the index becomes to ensure that it doesn't break.

Reply:
the answer is not to test it. you're testing an implementation detail. test the external behaviour only instead (as in, treat the list as a black box that only has `IsInAnyInterval`).

exposing is wrong design-wise as it breaks encapsulation. this is internal state that should not be visible to inheritors.